Pnpm migration

hashicc opened the pull request:
# Description
<!-- Add a brief description of changes here -->

<!-- Uncomment line below to manually add link to JIRA ticket -->
<!-- :tickets: [Jira ticket](https://hashicorp.atlassian.net/browse/JIRA_TICKET_NUMBER) -->

## Screenshots (if appropriate)

## How to Test
<!-- Add steps to test this change. Include any other necessary relevant links -->

## Checklist
<!-- strikethrough the checklist item that is not relevant to your change -->

- [ ] I have added before and after screenshots for UI changes
- [ ] I have added JSON response output for API changes
- [ ] I have added steps to reproduce and test for bug fixes in the description
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works

ZedLi commented on ui/desktop/electron-app/package.json:
```diff
@@ -43,8 +43,16 @@
-    "highlight.js": "^10.4.1",
-    "node-gyp": "^10.0.0"
+  "pnpm": {
+    "onlyBuiltDependencies": [
```
Do we have to whitelist every dependency that has a native build/install script? When should we not include one here?

hashicc replied:
That's right, as security precaution it prevents compromised packages from getting to run `postinstall` changes without expecting and opting in. If one of those packages that was previously trusted gets compromised then you're in trouble but this is supposed to limit the blast radius.

> When should we not include one here?

Most packages should not be included. It [gives nice errors that let you know when something attempted to run](https://github.com/pnpm/pnpm/pull/8958/files#diff-439970ce8a3bde9ed7fa780c9bfa222d021210b9a608b0f72ab03a936902d812R90-R94) but was ignored and gives you the option to explicitly allow or ignore the package.

ZedLi reviewed (approved):
Tested admin/desktop builds and everything works well! Just left a few questions

hashicc commented on package.json:
```diff
@@ -61,6 +58,5 @@
   "engines": {
     "node": "20.* || 22.*"
-  },
-  "packageManager": "yarn@4.6.0"
```
That's a good question. It's kind of a magic field that seems like it's official and sort of was but now the [node technical steering committee](https://github.com/nodejs/TSC/pull/1697#issuecomment-2737093616) has voted to remove it from future releases of node. It'll still exist but it'll be less available and it's usefulness will be even less obvious.

I could add it back because it is supported by `pnpm/action-setup` but its sort of a magic link between setting it here and it getting picked up in CI. A small mention in the README could help remedy this.

Otherwise, removing it here, leaving the explicit pnpm version set in CI, and having a choose your own adventure for installation methods. `brew install pnpm` is probably the easiest way to go for the team or if people have other preferences. I like to use [`mise`](https://github.com/jdx/mise) for all my versioning and sometimes it adds a `.toml` in a project I'm working on (sort of like addition of `"packageManager"` when you used corepack) that I remove and just temporarily set the global version instead

aklkv replied:
I would advise to keep packageManager field and add instruction to manage pnpm version with core pack so you easily upgrade it with corepack up and CI will pick it up. We follow those tactics in cloud-ui, atlas, cloud-admin, hds. Tho node might be dropping corepack I am sure there will be alternatives either in the form of standalone corepack or something else, similarly to rome it's important to have single source of truth of pnpm version

hashicc commented on package.json:
```diff
@@ -45,14 +38,18 @@
+      "node-gyp": "^10.0.0",
+      "ansi-html": "^0.0.9",
+      "fireworm>async": "^2.6.4",
+      "ember-inflector": "^4.0.2"
```
This isn't necessary and I _think_ can be removed now. During the first pnpm spike before ember upgrades had settled the `ember-inflector` version was mismatched and embroider was calling this package out. I checked with `pnpm -r why ember-inflector` without the override and everything looks right

ZedLi commented on .github/workflows/build-admin-ui.yaml:
```diff
@@ -12,35 +12,34 @@ jobs:
-      - run: yarn run build:ui:admin:oss
+          cache: 'pnpm'
+          cache-dependency-path: '**/pnpm-lock.yaml'
+      - run: pnpm install --offline
```
For my own edification, what was the purpose of `pnpm install --offline`?

ZedLi reviewed (approved):
Thanks for this upgrade!! This looks fantastic! 🎉 🎊

DhariniJeeva reviewed (approved):
Thanks for this upgrade! 

amazing work!   🎉  

I just had one minor question- do we still need references to `yarn` in our gitignore files?

hashicc commented:
Thanks @DhariniJeeva and @lisbet-alvarez for catching those references. I've removed the yarn references from the *ignore files in https://github.com/hashicorp/boundary-ui/pull/2828/commits/e57557e56af95e91bb16b1d25e2c33574ca457c2, along with:
* npm shrinkwrap and lock files since we don't use `npm`
* `.travis.yml` since we don't use Travis CI

laurenolivia commented on README.md:
```diff
@@ -218,10 +218,10 @@ will not be accepted.
 [git]: https://git-scm.com/
 [node]: https://nodejs.org/
-[yarn]: https://yarnpkg.com/
+[pnpm]: https://pnpmpkg.com/
```
Got a server not found at this url. Should the link be: https://pnpm.io/ ?

laurenolivia commented on README.md:
```diff
@@ -218,10 +218,10 @@ will not be accepted.
 [chrome]: https://google.com/chrome/
 [firefox]: https://firefox.com/
-[yarn-workspaces]: https://yarnpkg.com/features/workspaces
+[pnpm-workspaces]: https://pnpmpkg.com/features/workspaces
```
Should this url be: https://pnpm.io/workspaces ?

lisbet-alvarez reviewed (approved):
Thank u for this big migration 🥳

cameronperera reviewed (approved):
Great work!

calcaide reviewed (dismissed):
Terrific work!! this is amazing how pnpm performs!!

I run some of our scripts such as format or lint successfully, in fact, not sure if introduced within this branch, but when running `pnpm format` under `ui/desktop` it changes 2 files 😄  ( `/desktop/tests/acceptance/authentication-test.js` and `/desktop/app/styles/app.scss`).

I also build and run admin and desktop with no problem 🙌 🚀 

Note: for the comments read, I installed [pnpm using corepack](https://pnpm.io/installation#using-corepack)

laurenolivia reviewed (commented):
Great job. Tested locally. Just found a couple more broken links.

hashicc commented on ui/desktop/README.md:
```diff
@@ -35,17 +35,17 @@ You will need the following things properly installed on your computer.
 - [Git](https://git-scm.com/)
 - [Node.js](https://nodejs.org/)
-- [Yarn](https://yarnpkg.com/)
+- [Pnpm](https://pnpmpkg.com/)
```
Thanks again!

hashicc commented:
Updated changes in https://github.com/hashicorp/boundary-ui/pull/2828/commits/78d97c9fbc711e984b84865f5d4f76847b6eeae0 look like they're working for the pnpm action in CI picking up the pnpm version to use from `packageManager`:
```
Run pnpm/action-setup@a7487c7e89a18df4991f7f222e4898a00d66ddda
  with:
    run_install: false
    dest: ~/setup-pnpm
    package_json_file: package.json
    standalone: false
Running self-installer...
  Progress: resolved 1, reused 0, downloaded 0, added 0
  Packages: +1
  +
  Progress: resolved 1, reused 0, downloaded 1, added 1, done
  
  dependencies:
  + pnpm 10.8.1 (10.11.1 is available)
  
  Done in 839ms
Installation Completed!]
```
